src/metabase/api/model_index.clj

```diff
@@ -0,0 +1,81 @@
+(ns metabase.api.model-index
+  (:require
+   [compojure.core :refer [POST]]
+   [metabase.api.common :as api]
+   [metabase.models.card :refer [Card]]
+   [metabase.models.model-index :as model-index :refer [ModelIndex]]
+   [metabase.task.index-values :as task.index-values]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+(defn- ensure-type
+  "Ensure that the ref exists and is of type required for indexing."
+  [t ref metadata]
+  (if-let [field (some (fn [f] (when ((comp #{(model-index/normalize-field-ref ref)} :field_ref) f)
```

Comment:
we have `qp.util/field->field-info` to find `field` in `result_metadata` given a `field_ref`,  there are edge cases where doing a simple equality check is not sufficient( I.e  aggregated fields or field_ref reference by name in native queries)

Reply:
I'll look into this, but i think we actually do want equality on the field ref. The ref is taken from the metadata so we don't need to do anything fancy to recognize it I think.

Reply:
that function actually tries doing an equality check first.

Then it'll fall back to check pair `name, base-type` to cover some edge cases for aggregation + column with id is a string.